src/model/interactions/message_component.rs

```diff
@@ -0,0 +1,689 @@
+                    .and_then(GuildId::deserialize)
+                    .map_err(DeError::custom)?,
+            ),
+            false => None,
+        };
+
+        let channel_id = map
+            .remove("channel_id")
+            .ok_or_else(|| DeError::custom("expected channel_id"))
+            .and_then(ChannelId::deserialize)
+            .map_err(DeError::custom)?;
+
+        let member = match map.contains_key("member") {
+            true => Some(
+                map.remove("member")
+                    .ok_or_else(|| DeError::custom("expected member"))
+                    .and_then(Member::deserialize)
+                    .map_err(DeError::custom)?,
+            ),
+            false => None,
+        };
+
+        let user = match map.contains_key("user") {
+            true => map
+                .remove("user")
+                .ok_or_else(|| DeError::custom("expected user"))
+                .and_then(User::deserialize)
+                .map_err(DeError::custom)?,
+            false => member.as_ref().expect("expected user or member").user.clone(),
+        };
+
+        let message = {
+            let message = map
+                .remove("message")
+                .ok_or_else(|| DeError::custom("expected message"))
+                .and_then(JsonMap::deserialize)
+                .map_err(DeError::custom)?;
+
+            let partial = !message.contains_key("author");
+
+            let value: Value = message.into();
+
+            if partial {
+                InteractionMessage::Ephemeral(
+                    EphemeralMessage::deserialize(value).map_err(DeError::custom)?,
+                )
+            } else {
+                InteractionMessage::Regular(Message::deserialize(value).map_err(DeError::custom)?)
+            }
+        };
+
+        let token = map
+            .remove("token")
+            .ok_or_else(|| DeError::custom("expected token"))
+            .and_then(String::deserialize)
+            .map_err(DeError::custom)?;
+
+        let version = map
+            .remove("version")
+            .ok_or_else(|| DeError::custom("expected version"))
+            .and_then(u8::deserialize)
+            .map_err(DeError::custom)?;
+
+        Ok(Self {
+            id,
+            application_id,
+            kind,
+            data,
+            message,
+            guild_id,
+            channel_id,
+            member,
+            user,
+            token,
+            version,
+        })
+    }
+}
+
+/// A message component interaction data, provided by [`MessageComponentInteraction::data`]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[non_exhaustive]
+pub struct MessageComponentInteractionData {
+    /// The custom id of the component.
+    pub custom_id: String,
+    /// The type of the component.
+    pub component_type: ComponentType,
+    /// The given values of the [`SelectMenu`]s
+    #[serde(default)]
+    pub values: Vec<String>,
+}
+
+// A component.
```

Comment:
@HarmoGlace, is there any reason why you kept the `Component` enum?
It seems to be no longer used beside the `From`/`TryFrom` implementations after the `Message::components` type  was changed to `Vec<ActionRow>`.

Reply:
Indeed, it is no longer used in the library itself, but as the `ComponentType` structure still exists, I thought that it might be consistent to also add a wrapper to any component. Feel free to make a PR to remove it, it wouldn't bother me

Reply:
What is the use case you had in mind?

Reply:
I implemented it for consistency : an enum for interaction exists, so an enum for each component might be interesting. It might be considered as a "util" structure.